File: src/utils/errorHandlers.ts
/**
 * Centralized error handling utilities
 */

export class DatagroomError extends Error {
  constructor(
    message: string,
    public code: string,
    public statusCode: number = 500
  ) {
    super(message);
    this.name = 'DatagroomError';
  }
}

export class DatasetNotFoundError extends DatagroomError {
  constructor(datasetName: string) {
    super(
      `Dataset '${datasetName}' not found`,
      'DATASET_NOT_FOUND',
      404
    );
    this.name = 'DatasetNotFoundError';
  }
}

export class InvalidFilterError extends DatagroomError {
  constructor(message: string) {
    super(
      `Invalid filter: ${message}`,
      'INVALID_FILTER',
      400
    );
    this.name = 'InvalidFilterError';
  }
}

export class FieldNotFoundError extends DatagroomError {
  constructor(fieldName: string, datasetName: string) {
    super(
      `Field '${fieldName}' does not exist in dataset '${datasetName}'`,
      'FIELD_NOT_FOUND',
      400
    );
    this.name = 'FieldNotFoundError';
  }
}

export class DatabaseConnectionError extends DatagroomError {
  constructor(message: string) {
    super(
      `Unable to connect to database: ${message}`,
      'DATABASE_CONNECTION_ERROR',
      503
    );
    this.name = 'DatabaseConnectionError';
  }
}

/**
 * Format error for MCP tool response
 */
export function formatError(error: unknown): string {
  if (error instanceof DatagroomError) {
    return `${error.name}: ${error.message}`;
  }
  
  if (error instanceof Error) {
    return `Error: ${error.message}`;
  }
  
  return `Unknown error: ${String(error)}`;
}

File: utils/error_handlers.py
"""
Centralized error handling utilities (matches TS errorHandlers.ts semantics).
"""

from typing import Any


class DatagroomError(Exception):
    """Base error for Datagroom MCP Server."""

    def __init__(
        self,
        message: str,
        code: str = "DATAGROOM_ERROR",
        status_code: int = 500,
    ):
        super().__init__(message)
        self.message = message
        self.code = code
        self.status_code = status_code
        self.name = "DatagroomError"


class DatasetNotFoundError(DatagroomError):
    def __init__(self, dataset_name: str):
        super().__init__(
            f"Dataset '{dataset_name}' not found",
            code="DATASET_NOT_FOUND",
            status_code=404,
        )
        self.name = "DatasetNotFoundError"


class InvalidFilterError(DatagroomError):
    def __init__(self, message: str):
        super().__init__(
            f"Invalid filter: {message}",
            code="INVALID_FILTER",
            status_code=400,
        )
        self.name = "InvalidFilterError"


class FieldNotFoundError(DatagroomError):
    def __init__(self, field_name: str, dataset_name: str):
        super().__init__(
            f"Field '{field_name}' does not exist in dataset '{dataset_name}'",
            code="FIELD_NOT_FOUND",
            status_code=400,
        )
        self.name = "FieldNotFoundError"


class DatabaseConnectionError(DatagroomError):
    def __init__(self, message: str):
        super().__init__(
            f"Unable to connect to database: {message}",
            code="DATABASE_CONNECTION_ERROR",
            status_code=503,
        )
        self.name = "DatabaseConnectionError"


def format_error(error: BaseException | Any) -> str:
    """Format error for MCP tool response (matches TS formatError)."""
    if isinstance(error, DatagroomError):
        return f"{error.name}: {error.message}"
    if isinstance(error, Exception):
        return f"Error: {error}"
    return f"Unknown error: {str(error)}"
